metacat-main/src/main/java/com/netflix/metacat/main/services/CatalogTraversal.java

```diff
@@ -0,0 +1,463 @@
+        @Nonnull @NonNull final Registry registry
+    ) {
+        this.config = config;
+        this.actions = Lists.newArrayList();
+        this.catalogService = catalogService;
+        this.databaseService = databaseService;
+        this.tableService = tableService;
+        this.registry = registry;
+    }
+
+    private static ExecutorService newFixedThreadPool(
+        final int nThreads,
+        final String threadFactoryName,
+        final int queueSize
+    ) {
+        return new ThreadPoolExecutor(nThreads, nThreads,
+            0L, TimeUnit.MILLISECONDS,
+            new LinkedBlockingQueue<>(queueSize),
+            new ThreadFactoryBuilder()
+                .setNameFormat(threadFactoryName)
+                .build(),
+            (r, executor) -> {
+                // this will block if the queue is full
+                try {
+                    executor.getQueue().put(r);
+                } catch (InterruptedException e) {
+                    throw Throwables.propagate(e);
+                }
+            });
+    }
+
+    /**
+     * Adds the action handlers.
+     * @param actionHandlers list of action handlers
+     */
+    public void addActions(final List<CatalogTraversalAction> actionHandlers) {
+        this.actions.addAll(actionHandlers);
+    }
+
+    /**
+     * Does a sweep across all catalogs to refresh the same data in elastic search.
+     */
+    public void process() {
+        processCatalogs(getCatalogNames());
+    }
+
+    /**
+     * Does a sweep across given catalogs to refresh the same data in elastic search.
+     *
+     * @param catalogNames catalog anmes
+     */
+    public void processCatalogs(final List<String> catalogNames) {
+        final List<QualifiedName> qNames = catalogNames.stream()
+            .map(QualifiedName::ofCatalog).collect(Collectors.toList());
+        _process(qNames, () -> _processCatalogs(catalogNames), "processCatalogs", true, 1000);
+    }
+
+    /**
+     * Does a sweep across given catalog and databases to refresh the same data in elastic search.
+     *
+     * @param catalogName   catalog
+     * @param databaseNames database names
+     */
+    public void processDatabases(final String catalogName, final List<String> databaseNames) {
+        final List<QualifiedName> qNames = databaseNames.stream()
+            .map(s -> QualifiedName.ofDatabase(catalogName, s)).collect(Collectors.toList());
+        _process(qNames, () -> _processDatabases(QualifiedName.ofCatalog(catalogName), qNames), "processDatabases",
+            true, 1000);
+    }
+
+    @SuppressWarnings("checkstyle:methodname")
+    private void _process(final List<QualifiedName> qNames, final Supplier<ListenableFuture<Void>> supplier,
+                          final String requestName, final boolean delete, final int queueSize) {
+        if (isTraversalAlreadyRunning.compareAndSet(false, true)) {
+            final long start = registry.clock().wallTime();
```

Comment:
How long will this traversal run? If it's a very long running process, we should set the lock at the individual catalog level.

Reply:
At present, I am not changing the behavior. We should revisit this later.